cleanlab/outlier.py

```diff
@@ -470,7 +455,16 @@ def _get_ood_features_scores(
         # Get distances to k-nearest neighbors Note that the knn object contains the specification of distance metric
         # and n_neighbors (k value) If our query set of features matches the training set used to fit knn, the nearest
         # neighbor of each point is the point itself, at a distance of zero.
-        distances, _ = knn.kneighbors(features)
+        distances, indices = knn.kneighbors(features)
+        if correct_knn:  # This should only happen if knn is None at the start of this function
+            _features_for_correction = (
+                knn._fit_X if features is None else features
+            )  # Hacky way to get features (training or test). Storing np.unique results is a hassle.
+            distances, _ = correct_knn_distances_and_indices(
+                features=_features_for_correction,
+                distances=distances,
+                indices=indices,
+            )
```

Comment:
Why isn't the correction happening inside of the helper method: `features_to_knn()` or `construct_knn_graph_from_index()` ?  Shouldnt the correction be happening inside of these helpers, so that this correction consistently happens anytime the KNN graph is constructed?

Consider what happens if user runs Datalab sequentially one issue-type at a time. Say they do for one Datalab in order A: outliers, duplicates, non-iid

Then they do for another Datalab in order B: duplicates, non-iid, outliers. 

Isn't it important that the KNN construction and data point distances are consistently represented, so that the end behavior is correct regardless of order A vs B?

Reply:
or are you saying here it has to be done this way to support test data?
Not sure why that'd be, since for test data we'd never be in situation where `knn` is None, right

Reply:
Accessing `knn._fit_X` directly is a short-term solution, but moving the correction logic inside `construct_knn_graph_from_index(knn: Nearestneighbors)` may be acceptable for now.  I'd suggest passing `correct_knn` as a parameter to make this correction opt-in, ensuring it only happens when necessary, such as when `features` are missing but need correction via the `knn` object.

Reply:
yes I think the logic needs to be moved inside the method, otherwise it seems too bug prone as users call Datalab in different issue-type orders and in different ways. Having an opt-in arg for the method seems fine

Reply:
Logic is now in the helper functions, but the code in `cleanlab/outlier.py` is an exception, as that code is fundamentally different from other places in the code-base that rely on a `knn_graph: csr_matrix`.

The outlier issue check in `Datalab`  handle's this correctly and is the recommended way for handling outliers efficiently, even with corrections for exact duplicates.